### PR TITLE
Attempt to unrevert: Skip over tag attributes when selecting inside tags

### LINKED ISSUE #117: Select inside brackets for HTML files doesn't consider HTML attributes
Originally reported over at https://github.com/atom/atom/issues/6008 (thanks @philsinatra!).

Example:

``` html
<div class="wrapper">
  <div class="main">
    <p>Lorem ipsum dolor sit amet, consectetur adipisicing elit. Corrupti ipsam ullam magni minus, aperiam animi, quasi corporis. Impedit expedita, excepturi, iure corporis quo maxime, aperiam perferendis voluptas praesentium sint quam.</p>
  </div>
</div>
```

Position the cursor after `class="wrapper">` and execute the "Select inside brackets" command. Result:

screen shot 2015-03-18 at 18 43 34

Instead, I expected to see this:

screen shot 2015-03-18 at 18 52 37

I'm guessing that the command doesn't expect it will see any attributes, so it just selects starting from where it thinks the `>` of the opening tag will be?

### PR DESCRIPTION
**Pull request by ungb**
NOTE: this was an attempt to cherry pick changes from https://github.com/atom/bracket-matcher/pull/302 since I had to revert these changes previous to revert a previous change.  

### Requirements

* Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* All new code requires tests to ensure against regressions

### Description of the Change

When selecting inside tags, grab the full range of the tag rather than just the tag name range.  This change allows the attributes to be skipped, leading to the actual area between tags being highlighted.

### Alternate Designs

None.

### Benefits

Using the `bracket-matcher:select-inside-brackets` command will work correctly for tags with attributes.

### Possible Drawbacks

There may be a slight performance decrease compared to before when the cursor is currently on a tag.  This is because we have to re-search for the tags in order to get the full range (i.e. there is no way that I know of to grab the full range from the existing range).

### Applicable Issues

Fixes #117

/cc @50Wliu @Ben3eeE

### CODE REVIEW

File: lib/tag-finder.coffee [hunk 1]
```diff
@@ -85,17 +85,28 @@ class TagFinder
...
+            startRange.start = startRange.start.translate([0, 1])
+
+            # End right after the tag name
+            startRange.end = startRange.start.translate([0, tagName.length])
+
+            #TODO: this was originial change in PR that doesn't work.
+            # Subtract < and tag name suffix from range
+            # startRange = range.translate([0, 1], [0, -(match[2].length + match[3].length)])
+
+          until startRange.end.column > 0 # if the tag spans multiple lines, the end.column is negative, so iterate until it's not.
```
Comment (50Wliu) [inline lib/tag-finder.coffee]: Not sure this is needed, I think it was included in #257 but not in #304.

File: lib/tag-finder.coffee [hunk 2]
```diff
@@ -134,19 +146,23 @@ class TagFinder
...
 
       ranges = {startRange, endRange} if startRange? and endRange?
     ranges
 
-  findEnclosingTags: ->
-    if ranges = @findStartEndTags()
+  findEnclosingTags: (fullRange=false) ->
+    if ranges = @findStartEndTags(fullRange)
       if @isTagRange(ranges.startRange) and @isTagRange(ranges.endRange)
+        if ranges.startRange.start.row > ranges.endRange.start.row
```
Comment (50Wliu) [inline lib/tag-finder.coffee]: Ditto

**Comment by Ben3eeE** [comment]
Now it's just the specs for `bracket-match:select-inside-brackets` that are failing 🎉

File: lib/tag-finder.coffee [hunk 3]
```diff
@@ -85,17 +85,25 @@ class TagFinder
...
+          unless fullRange
+            # Move the start past the initial <
+            startRange.start = startRange.start.translate([0, 1])
+
+            # End right after the tag name
+            startRange.end = startRange.start.translate([0, tagName.length])
+
+            #TODO: this was originial change in PR that doesn't work.
+            # Subtract < and tag name suffix from range
+            # startRange = range.translate([0, 1], [0, -(match[2].length + match[3].length)])
```
Comment (50Wliu) [inline lib/tag-finder.coffee]: Try
```coffee
startRange.start = range.start.translate([0, 1])
startRange.end = range.end.translate([0, -match[3].length]) if match[3]?
startRange.end = range.end.translate([0, -match[4].length])
```
Might want to clean that up if that works.

**Comment by Ben3eeE** [comment]
@ungb The specs seems to be passing now but there is a hang from these changes that I discovered that we need to investigate before merging

**Comment by 50Wliu** [comment]
`endPosition = @editor.getLastCursor().getCurrentWordBufferRange({wordRegex: @endOfTagRegex}).end`.  I believe that change was from #304.

**Comment by Ben3eeE** [comment]
This looks like it's ready to go. I can test it out after https://github.com/atom/bracket-matcher/issues/314 has been fixed so we can release both this and the self closing tags when we bump it on atom/atom.

**Comment by ungb** [comment]
@50Wliu @Ben3eeE I've merged https://github.com/atom/bracket-matcher/pull/315, Are you guys ok with me merging this change after rerunning test?

**Comment by Ben3eeE** [comment]
@ungb I want to run some manual testing on this to ensure no regressions before merging.

**Comment by Ben3eeE** [comment]
@50Wliu I'm seeing an exception

**Atom**: 1.20.0 x64
**Electron**: 1.6.9
**OS**: Microsoft Windows 10 Pro
**Thrown From**: [bracket-matcher](https://github.com/atom/bracket-matcher) package 0.87.3


### Stack Trace

Uncaught TypeError: Cannot read property 'startRange' of null

```
At I:\atom_master\bracket-matcher\lib\bracket-matcher-view.coffee:271

TypeError: Cannot read property 'startRange' of null
    at BracketMatcherView.module.exports.BracketMatcherView.selectInsidePair (I:/atom_master/bracket-matcher/lib/bracket-matcher-view.coffee:271:10)
    at HTMLElement.<anonymous> (I:/atom_master/bracket-matcher/lib/bracket-matcher-view.coffee:36:8)
    at CommandRegistry.module.exports.CommandRegistry.handleCommandEvent (~/AppData/Local/atom/app-1.20.0/resources/app/src/command-registry.js:265:35)
    at ~/AppData/Local/atom/app-1.20.0/resources/app/src/command-registry.js:3:65
    at Object.didConfirmSelection (~/AppData/Local/atom/app-1.20.0/resources/app/node_modules/command-palette/lib/command-palette-view.js:106:35)
    at SelectListView.confirmSelection (~/AppData/Local/atom/app-1.20.0/resources/app/node_modules/atom-select-list/src/select-list-view.js:313:26)
    at HTMLDivElement.core:confirm (~/AppData/Local/atom/app-1.20.0/resources/app/node_modules/atom-select-list/src/select-list-view.js:81:20)
    at CommandRegistry.module.exports.CommandRegistry.handleCommandEvent (~/AppData/Local/atom/app-1.20.0/resources/app/src/command-registry.js:265:35)
    at ~/AppData/Local/atom/app-1.20.0/resources/app/src/command-registry.js:3:65
    at KeymapManager.module.exports.KeymapManager.dispatchCommandEvent (~/AppData/Local/atom/app-1.20.0/resources/app/node_modules/atom-keymap/lib/keymap-manager.js:621:22)
    at KeymapManager.module.exports.KeymapManager.handleKeyboardEvent (~/AppData/Local/atom/app-1.20.0/resources/app/node_modules/atom-keymap/lib/keymap-manager.js:412:28)
    at WindowEventHandler.module.exports.WindowEventHandler.handleDocumentKeyEvent (~/AppData/Local/atom/app-1.20.0/resources/app/src/window-event-handler.js:100:42)
    at HTMLDocument.<anonymous> (~/AppData/Local/atom/app-1.20.0/resources/app/src/window-event-handler.js:3:65)
```

### Commands

```
     -0:23.3.0 core:delete (input.hidden-input)
 14x -0:22.4.0 core:move-right (input.hidden-input)
     -0:21.4.0 intentions:highlight (input.hidden-input)
     -0:21.1.0 command-palette:toggle (input.hidden-input)
     -0:20.3.0 core:confirm (input.hidden-input)
     -0:20.3.0 bracket-matcher:select-inside-brackets (input.hidden-input)
     -0:19.2.0 command-palette:toggle (input.hidden-input)
     -0:18.7.0 core:confirm (input.hidden-input)
     -0:18.7.0 bracket-matcher:select-inside-brackets (input.hidden-input)
     -0:14.8.0 command-palette:toggle (input.hidden-input)
     -0:14.5.0 core:confirm (input.hidden-input)
     -0:14.5.0 bracket-matcher:select-inside-brackets (input.hidden-input)
     -0:08.7.0 intentions:highlight (input.hidden-input)
     -0:08.6.0 command-palette:toggle (input.hidden-input)
     -0:08.3.0 core:confirm (input.hidden-input)
     -0:08.3.0 bracket-matcher:select-inside-brackets (input.hidden-input)
```

### Non-Core Packages

```
advanced-open-file 0.16.6 
atom-beautify 0.30.5 
atom-dark-fusion-syntax 2.2.0 
atom-ide-ui 0.1.11 
atom-material-syntax 1.0.6 
atom-material-ui undefined 
autocomplete-emojis 2.5.0 
busy-signal 1.4.3 
chester-atom-syntax 0.3.0 
chestnut-light-atom-syntax 0.2.1 
color-picker 2.2.5 
editor-stats 0.17.0 
fizzy 0.21.0 
golden-ratio 0.3.0 
hey-pane 1.0.0 
highlight-selected 0.13.1 
honor-altgraph 0.1.0 
intentions 1.1.5 
language-latex 1.1.1 
language-markdown 0.25.1 
latex 0.46.0 
less-than-slash 0.17.0 
linter 2.2.0 
linter-js-standard 4.0.0 
linter-ui-default 1.6.8 
markdown-pdf 1.5.4 
minimap 4.29.6 
minimap-autohider 1.5.3 
minimap-find-and-replace 4.5.2 
minimap-highlight-selected 4.6.1 
nebula-syntax 0.4.4 
nebula-ui 0.6.0 
no-caffeine-syntax 0.18.1 
nucleus-dark-ui 0.12.3 
pastel-jelly-syntax 0.2.0 
php-cs-fixer 4.1.0 
pigments 0.40.2 
platformio-ide-terminal 2.5.5 
preview-inline 1.4.7 
remote-edit 1.9.0 
seti-syntax 1.1.3 
seti-ui 1.9.0 
Sublime-Style-Column-Selection 1.7.4 
tab-title 0.3.5 
toggler 0.3.0 
vim-mode-plus 1.3.3 
zentabs 0.8.8 
```

**Comment by ungb** [comment]
Wonder if this could of been caused by me merging master into this.

**Comment by Ben3eeE** [comment]
I can reproduce it with the following:

```html
<body class='markdown-preview' data-use-github-style>
    <span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span><span></span></body>
```

And placing the cursor on the end of the first line and running `bracket-matcher:select-inside-brackets`.

**Comment by Ben3eeE** [comment]
@50Wliu @ungb This looks good now 👍 🚢

**Pull request merged**